### PR TITLE
Experimental Support for Android (x86_64 and AArch64)

### PR DESCRIPTION
**Pull request by igrep**
# Description

Related: https://github.com/wasmerio/wasmer/issues/1113

# Current Status

- ~~This pull request is a draft until the next version of libc crate (which should contain https://github.com/rust-lang/libc/pull/1622)~~
    - Now released! 🎉
- I confirmed the tests of wasmer-runtime-core pass on Android x86\_64.
    - The other tests including ones on AArch64 seems too hard so far... 😓 

# Review

- [x] Add a short description of the the change to the CHANGELOG.md file

# Note

I'm happy given any hint to test of Android specific code (esp. in libc I added in https://github.com/rust-lang/libc/pull/1622).

### CODE REVIEW

**Comment by igrep** [comment]
> - I confirmed the tests of wasmer-runtime-core pass on Android x86\_64.
    - The other tests including ones on AArch64 seems too hard so far... 😓 

I've also pushed the test scripts in [another branch](https://github.com/igrep/wasmer/commit/0c5c827e6df2ad10f84c614ab1242c4bd15ecf1e).
Can I add them to the CI workflow?

**Comment by syrusakbary** [comment]
Awesome! 

Let's add this change into the CHANGELOG.md, so once libc publishes the new version we can merge it :)

**Comment by syrusakbary** [comment]
> Can I add them to the CI pipelines?

We can merge the android tests in a separate PR. @igrep do you have any insight in how long take the tests to run with Android?

**Comment by igrep** [comment]
Here's the result of `time` of the test script on my laptop after `cargo clean` and adding `--no-cache` option to the docker command in the script:

```
> time ci/run-docker.sh x86_64-linux-android --manifest-path=lib/runtime-core-tests/Cargo.toml &> wasmer-test-result.txt

real    11m49.587s
user    0m0.340s
sys     0m0.355s
```

The test itself is fast enough: Most of the time is building the docker image.

File: lib/singlepass-backend/src/lib.rs [hunk 1]
```diff
@@ -17,8 +17,10 @@
     all(target_os = "macos", target_arch = "x86_64"),
     all(target_os = "linux", target_arch = "x86_64"),
     all(target_os = "linux", target_arch = "aarch64"),
+    all(target_os = "android", target_arch = "x86_64"),
+    all(target_os = "android", target_arch = "aarch64"),
 )))]
-compile_error!("This crate doesn't yet support compiling on operating systems other than FreeBSD, linux and macos and architectures other than x86_64");
+compile_error!("This crate doesn't yet support compiling on operating systems and architectures other than these:\n       - FreeBSD and x86_64\n       - FreeBSD and AArch64\n       - macOS and x86_64\n       - Linux and x86_64\n       - Linux and AArch64\n       - Android and x86_64\n       - Android and AArch64");
```
Comment (MarkMcCaskey) [inline lib/singlepass-backend/src/lib.rs]: I'd prefer a multi-line string here for easier editing, reading, and verifying that the indentation is correct!

```Rust
"This crate doesn't yet support compiling on operating systems and architectures other than these:
       - FreeBSD and x86_64
       - FreeBSD and AArch64
       - macOS and x86_64
       - Linux and x86_64
       - Linux and AArch64
       - Android and x86_64
       - Android and AArch64"
```

File: Cargo.toml [hunk 2]
```diff
@@ -116,3 +116,7 @@ crate-type = ["bin"]
 [[example]]
 name = "callback"
 crate-type = ["bin"]
+
+[patch.crates-io]
+libc = { git = "https://github.com/rust-lang/libc.git" }
```
Comment (Hywan) [inline Cargo.toml]: Why is this required?

Comment (igrep) [reply Cargo.toml]: Without this line, packages depending on libc tries to use the released version of libc.
This line https://github.com/wasmerio/wasmer/pull/1292/files#diff-1cb0a48640bca2c4affbcf2f5e9a81c9R20 wasn't enough

Anyway I'll delete this line after

File: Cargo.toml [hunk 3]
```diff
@@ -116,3 +116,6 @@ crate-type = ["bin"]
 [[example]]
 name = "callback"
 crate-type = ["bin"]
+
+[patch.crates-io]
+wabt-sys = { git = "https://github.com/pepyakin/wabt-rs.git" }
```
Comment (syrusakbary) [inline Cargo.toml]: Just created this issue in WABT to publish a new version in crates.io:
https://github.com/pepyakin/wabt-rs/issues/65

Once is merged we should be ready to go!

Comment (pepyakin) [reply Cargo.toml]: wabt-sys 0.7.1 is out.

Comment (syrusakbary) [reply Cargo.toml]: @igrep as soon as this line is updated we will merge! 🎉

Thanks @pepyakin for the quick release!

**Comment by syrusakbary** [comment]
I'll merge the PR and do extra work in a PR apart

**Comment by igrep** [comment]
I hasten to tell an issue before merging: I found a stack overflow error when running a simple WASI executable ("Hello, world" in Rust) on my Android phone (aarch64), not on an emulator (x86_64)

**Comment by igrep** [comment]
One more note: the first error reported by https://github.com/wasmerio/wasmer/pull/1292#issuecomment-600885850 should have been fixed.

**Comment by syrusakbary** [comment]
bors try

**Comment by igrep** [comment]
bors try

**Comment by MarkMcCaskey** [comment]
bors try

**Comment by syrusakbary** [comment]
Somehow test didn't run ?
Let's try again:

bors try

**Comment by syrusakbary** [comment]
bors try-

**Comment by syrusakbary** [comment]
bors try

**Comment by syrusakbary** [comment]
We should be good to merge once we solve the conflict @igrep!

**Pull request merged**